### PR TITLE
Pr mag cal fix

### PR DESCRIPTION
**Pull request by mhkabir**
This PR : 

- Fixes a bad assumption in sensor calibration code, which assumed that uORB instance ID == the ID of the file descriptor (e.g /dev/sensor_gyroX) for a sensor. This would lead to offsets being applied to the wrong sensor depending on the race between driver instantiation and uORB advertisement. This was hidden so far because of the startup order of sensors on non-v4Pro and non-v5 boards.
- Enables temperature compensation and boot-calibration for HMC5x83 mags running on v4Pro and v5 hardware, which should improve performance in general.
- Fixes MPU9250 mag driver to properly publish it's device ID. 
- Fixes LIS3MDL mag driver to properly publish it's device ID. 
- Adds checks to the sensor calibration code which will fail the calibration if there are inconsistencies in device IDs (usually pointing to a buggy sensor driver) so that we can catch them out and fix them. These problems have remained hidden so far because nothing was forcing a sanity check.

This needs : 

- Test flights on Pixhawk 1 / Pixhawk Mini, Pixhawk 2.1, Pixracer and Pixhawk Pro. Full parameter reset and redo all calibrations before flying.
- Bench tests on Snapdragon Flight / Bebop / RPi to make sure that calibration still works as normal.

### CODE REVIEW

**Comment by LorenzMeier** [comment]
Awesome work!

File: src/modules/commander/accelerometer_calibration.cpp [hunk 1]
```diff
@@ -465,36 +465,64 @@ calibrate_return do_accel_calibration_measurements(orb_advert_t *mavlink_log_pub
...
+			if(report.device_id == device_id[cur_accel]) {
+				// Device IDs match, correct ORB instance for this accel
+				found_cur_accel = true;
+				// store initial timestamp - used to infer which sensors are active
+				timestamps[cur_accel] = report.timestamp;
+			} else {
+				orb_unsubscribe(worker_data.subs[cur_accel]);
+			}
+
+#elif defined(__PX4_QURT) || defined(__PX4_POSIX_RPI) || defined(__PX4_POSIX_BEBOP)
```
Comment (bkueng) [inline src/modules/commander/accelerometer_calibration.cpp]: This can be simplified to `#else`

**Comment by mhkabir** [comment]
We still need the fix is more places, I think. Flight test on Pixhawk Pro with correct mag orientation showed toiletbowling, so it seems that there is something still wrong. Will keep working on this.

**Comment by LorenzMeier** [comment]
Can you please share a log? I don't think that validating it using a flight says anything about correctness - it just says that the system you flew wasn't right.

It could be that the calibration data was wrongly assigned and is now correctly, etc. Please try to check separately for correctness and we'll handle any churn (e.g. a required re-calibration) separately iff required.

**Comment by mhkabir** [comment]
More locations in the code needed the fix. See the commit I just pushed : https://github.com/PX4/Firmware/pull/7287/commits/6f8997ea5d6d7f56e6336b5b207438f7cf8cbb40

However, since now we only set the rotation matrix on startup, it would require a reboot after calibration. I added the `@reboot_required` flags to the params, but this might need some more support on QGC calibrator side to properly propagate up to the user.

File: src/modules/commander/accelerometer_calibration.cpp (hunk 1 above)
Comment (mhkabir) [inline src/modules/commander/accelerometer_calibration.cpp]: Done. Also extended to include Excelsior boards. 

Side note : These defines are yuck.

Comment (bkueng) [inline src/modules/commander/accelerometer_calibration.cpp]: How about  `#ifdef __PX4_NUTTX`?

Comment (mhkabir) [inline src/modules/commander/accelerometer_calibration.cpp]: Done. :)

**Comment by Avysuarez** [comment]
pixhawk mini; good flight, no issues.
http://review.px4.io/plot_app?log=f7047899-fd47-4ec6-9897-1b50642af153

**Comment by mhkabir** [comment]
OK, looks good. Can you test on v4 hardware? Normally it would not tell us much because v1-v4 hardware didn't expose the bug, but I will give you a custom binary in that case, which will force that situation. It would be good to fly with the board **and** external compass rotated (on yaw) with respect to the frame (any rotation should do). Make sure you reset params and recalibrate, setting the correct orientations during calibration. 

Here is the binary for Pixracer : 
[px4fmu-v4_default.px4.zip](https://github.com/PX4/Firmware/files/1032331/px4fmu-v4_default.px4.zip)

**Comment by santiago3dr** [comment]
flight on pixracer
http://logs.px4.io/plot_app?log=c7c761de-baca-4b8b-974a-8aa984768e59

**Comment by santiago3dr** [comment]
couple successful flight with the pixracer; external mag rotation set to yaw 270

http://logs.px4.io/plot_app?log=56252cbf-206b-43ca-8ea7-0e3129a70a7f
http://logs.px4.io/plot_app?log=770f0776-ced8-4a82-8217-80c72db9cb4f

**Comment by mhkabir** [comment]
@santiago3dr @Avysuarez Thanks for the testing. Looks good!

@LorenzMeier Good to merge after CI passes, and we make sure that QGC asks for a reboot after setting board/compass rotation.

**Comment by mhkabir** [comment]
@jywilson @mcharleb Can you please test this on Snapdragon? Please do a parameter reset and recalibrate all sensors to make sure that the calibration routine still functions properly. A bench test will be enough.

**Pull request merged**